subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java

```diff
@@ -43,8 +47,8 @@
 public class DefaultTaskOutputs implements TaskOutputsInternal {
     private final FileCollection allOutputFiles;
     private AndSpec<TaskInternal> upToDateSpec = AndSpec.empty();
-    private AndSpec<TaskInternal> cacheIfSpec = AndSpec.empty();
-    private OrSpec<TaskInternal> doNotCacheIfSpec = OrSpec.empty();
+    private List<SpecWithDescription<TaskInternal>> cacheIfSpecs = new ArrayList<SpecWithDescription<TaskInternal>>();
+    private List<SpecWithDescription<TaskInternal>> doNotCacheIfSpecs = new ArrayList<SpecWithDescription<TaskInternal>>();
```

Comment:
Maybe we can use this interface that Cedric's adding: https://github.com/gradle/gradle/pull/1178/files#diff-781b81824187b0f1558206a030f0c4dc ?